### PR TITLE
MAILBOX-321 PGP signature should always be considered as attachment

### PR DESCRIPTION
**Pull request by chibenwa**

### CODE REVIEW

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/mail/model/impl/MessageParser.java [hunk 1]
```diff
@@ -105,25 +110,28 @@ private boolean isMessageWithOnlyOneAttachment(Optional<ContentDispositionField>
         return contentDisposition.isPresent() && contentDisposition.get().isAttachment();
     }
 
-    private List<MessageAttachment> listAttachments(Multipart multipart, Context context) throws IOException {
-        ImmutableList.Builder<MessageAttachment> attachments = ImmutableList.builder();
+    private Stream<MessageAttachment> listAttachments(Multipart multipart, Context context) throws IOException {
+        ThrowingFunction<Entity, Stream<MessageAttachment>> listAttachments = Throwing.function(entity -> listAttachments(entity, context));
```
Comment (mbaechler) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/mail/model/impl/MessageParser.java]: you don't need to do that, you can just remove IOException from listAttachments signature

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/mail/model/impl/MessageParser.java [hunk 2]
```diff
@@ -73,17 +73,18 @@
     private static final Logger LOGGER = LoggerFactory.getLogger(MessageParser.class);
 
     private final Cid.CidParser cidParser;
+    private final DefaultMessageBuilder defaultMessageBuilder;
 
     public MessageParser() {
         cidParser = Cid.parser()
             .relaxed()
             .unwrap();
+        defaultMessageBuilder = new DefaultMessageBuilder();
```
Comment (mbaechler) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/mail/model/impl/MessageParser.java]: i'm not comfortable with that kind of change, DefaultMessageBuilder happens to (likely) be thread safe and only stores configuration as state but it's not even in its documentation.
Do we gain anything with this change ? If we want to keep that, can we add documentation on DefaultMessageBuilder to express the thread-safe and stateless properties ?

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/mail/model/impl/MessageParser.java [hunk 3]
```diff
@@ -194,13 +195,24 @@ private boolean isInline(Optional<ContentDispositionField> contentDispositionFie
     }
 
     private boolean isAttachment(Entity part, Context context) {
-        if (context == Context.BODY && isTextPart(part)) {
+        boolean bodyTextCriterion = context == Context.BODY && isTextPart(part);
+
+        boolean attachmentDispositionCriterion = getContentDispositionField(part)
+            .map(ContentDispositionField::getDispositionType)
+            .map(dispositionType -> dispositionType.toLowerCase(Locale.US))
```
Comment (mbaechler) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/mail/model/impl/MessageParser.java]: why do we have to do string handling here ? Can't we rely on ContentDispositionField for that ?

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/mail/model/impl/MessageParser.java [hunk 4]
```diff
@@ -194,13 +195,24 @@ private boolean isInline(Optional<ContentDispositionField> contentDispositionFie
...
+            .map(ATTACHMENT_CONTENT_DISPOSITIONS::contains)
+            .orElse(false);
+
+        boolean attachmentContentTypeCriterion =  getContentTypeField(part)
+            .map(ContentTypeField::getMimeType)
```
Comment (mbaechler) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/mail/model/impl/MessageParser.java]: same question about string here

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/mail/model/impl/MessageParser.java [hunk 5]
```diff
@@ -194,13 +195,24 @@ private boolean isInline(Optional<ContentDispositionField> contentDispositionFie
...
+            .map(dispositionType -> dispositionType.toLowerCase(Locale.US))
+            .map(ATTACHMENT_CONTENT_TYPES::contains)
+            .orElse(false);
+
+        if (bodyTextCriterion) {
```
Comment (mbaechler) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/mail/model/impl/MessageParser.java]: what is the point of extracting it to a variable ?

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/mail/model/impl/MessageParser.java (hunk 2 above)
Comment (chibenwa) [reply mailbox/store/src/main/java/org/apache/james/mailbox/store/mail/model/impl/MessageParser.java]: I'll remove that commit

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/mail/model/impl/MessageParser.java (hunk 5 above)
Comment (chibenwa) [reply mailbox/store/src/main/java/org/apache/james/mailbox/store/mail/model/impl/MessageParser.java]: I found that it ease reading to extract all conditions affecticting wether or not a part is an attachment to give them a name.

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/mail/model/impl/MessageParser.java (hunk 3 above)
Comment (chibenwa) [reply mailbox/store/src/main/java/org/apache/james/mailbox/store/mail/model/impl/MessageParser.java]: That was done before.

Moreover I don't see why we should correlate external library changes to a fix.

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/mail/model/impl/MessageParser.java (hunk 4 above)
Comment (chibenwa) [reply mailbox/store/src/main/java/org/apache/james/mailbox/store/mail/model/impl/MessageParser.java]: Same thing

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/mail/model/impl/MessageParser.java (hunk 1 above)
Comment (chibenwa) [reply mailbox/store/src/main/java/org/apache/james/mailbox/store/mail/model/impl/MessageParser.java]: Well catched!!!

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/mail/model/impl/MessageParser.java [hunk 6]
```diff
@@ -70,6 +70,7 @@
     private static final List<String> ATTACHMENT_CONTENT_DISPOSITIONS = ImmutableList.of(
             ContentDispositionField.DISPOSITION_TYPE_ATTACHMENT.toLowerCase(Locale.US),
             ContentDispositionField.DISPOSITION_TYPE_INLINE.toLowerCase(Locale.US));
+    private static final ImmutableList<String> ATTACHMENT_CONTENT_TYPES = ImmutableList.of("application/pgp-signature");
```
Comment (rouazana) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/mail/model/impl/MessageParser.java]: s/ImmutableList/List/

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/mail/model/impl/MessageParser.java (hunk 3 above)
Comment (mbaechler) [reply mailbox/store/src/main/java/org/apache/james/mailbox/store/mail/model/impl/MessageParser.java]: 1. mime4j is part of james
2. I don't know ContentDispositionField, so I was only asking if it's doable
3. If we don't care about mime4j being a great lib, who will ? I'm sad when we fallback to javax.mail or workaround limitation

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/mail/model/impl/MessageParser.java [hunk 7]
```diff
@@ -208,7 +206,7 @@ private boolean isAttachment(Entity part, Context context) {
             .map(ATTACHMENT_CONTENT_TYPES::contains)
             .orElse(false);
 
-        if (bodyTextCriterion) {
+        if (context == Context.BODY && isTextPart(part)) {
```
Comment (rouazana) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/mail/model/impl/MessageParser.java]: this should go first, no?

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/mail/model/impl/MessageParser.java (hunk 5 above)
Comment (mbaechler) [reply mailbox/store/src/main/java/org/apache/james/mailbox/store/mail/model/impl/MessageParser.java]: you can extract it to a method then, that we you have naming

**Comment by chibenwa** [comment]
```
just a question: a missing content disposition does not mean it is an attachment?
```

Before: -> Yes, necessarilly

After => Not if it is a PGP signature ;-)

I wondered wether Content-Types conidered as attachment should be  a white list or a black list. White list was less destructive to the code base (I met some corner cases with some empty parts / some informative part that should not be displayed.

After all, I expect the **No ContentDisposition but hey, it should be an attachment** to be rare. And if it happens again, then the fix is very simple (add to the white list).

Also, initially, the definition of attachment **is the content disposition**. The point is: JMAP acts also as a part of MUA, hence  we end up with display issues. (and I don't want to be too far from original definition).

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/mail/model/impl/MessageParser.java (hunk 7 above)
Comment (chibenwa) [reply mailbox/store/src/main/java/org/apache/james/mailbox/store/mail/model/impl/MessageParser.java]: I wanted to have decisions grouped in the same part of the method to ease reading...

**Comment by rouazana** [comment]
> Before: -> Yes, necessarilly

After: gpgSignatureShouldBeConsideredAsAnAttachment

I'm missing something?

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/mail/model/impl/MessageParser.java (hunk 6 above)
Comment (mbaechler) [reply mailbox/store/src/main/java/org/apache/james/mailbox/store/mail/model/impl/MessageParser.java]: I don't see what it brings

Comment (rouazana) [reply mailbox/store/src/main/java/org/apache/james/mailbox/store/mail/model/impl/MessageParser.java]: nothing special here, but in my experience we tend to reuse them and then leak immutablelist where it's not significant

**Comment by chibenwa** [comment]
```
[0691d6906f8e0af00de8c8d20837dd34bfffb7e8] [ERROR] Command was /bin/sh -c cd /james-project/server/protocols/jmap-integration-testing/cassandra-jmap-integration-testing && /usr/lib/jvm/java-8-openjdk-amd64/jre/bin/java -Xms256m -Xmx512m -jar /james-project/server/protocols/jmap-integration-testing/cassandra-jmap-integration-testing/target/surefire/surefirebooter5455741819124337900.jar /james-project/server/protocols/jmap-integration-testing/cassandra-jmap-integration-testing/target/surefire 2017-11-28T19-52-53_490-jvmRun1 surefire8928533235117125955tmp surefire_7475257416831253784770tmp
[0691d6906f8e0af00de8c8d20837dd34bfffb7e8] [ERROR] Error occurred in starting fork, check output in log
[0691d6906f8e0af00de8c8d20837dd34bfffb7e8] [ERROR] Process Exit Code: 1
[0691d6906f8e0af00de8c8d20837dd34bfffb7e8] [ERROR] Crashed tests:
[0691d6906f8e0af00de8c8d20837dd34bfffb7e8] [ERROR] org.apache.james.jmap.cassandra.CassandraSetMessagesMethodTest
[0691d6906f8e0af00de8c8d20837dd34bfffb7e8] [ERROR] 	at org.apache.maven.plugin.surefire.booterclient.ForkStarter.fork(ForkStarter.java:679)
[0691d6906f8e0af00de8c8d20837dd34bfffb7e8] [ERROR] 	at org.apache.maven.plugin.surefire.booterclient.ForkStarter.fork(ForkStarter.java:533)
[0691d6906f8e0af00de8c8d20837dd34bfffb7e8] [ERROR] 	at org.apache.maven.plugin.surefire.booterclient.ForkStarter.access$600(ForkStarter.java:117)
[0691d6906f8e0af00de8c8d20837dd34bfffb7e8] [ERROR] 	at org.apache.maven.plugin.surefire.booterclient.ForkStarter$2.call(ForkStarter.java:429)
[0691d6906f8e0af00de8c8d20837dd34bfffb7e8] [ERROR] 	at org.apache.maven.plugin.surefire.booterclient.ForkStarter$2.call(ForkStarter.java:406)
[0691d6906f8e0af00de8c8d20837dd34bfffb7e8] [ERROR] 	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
[0691d6906f8e0af00de8c8d20837dd34bfffb7e8] [ERROR] 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
[0691d6906f8e0af00de8c8d20837dd34bfffb7e8] [ERROR] 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
[0691d6906f8e0af00de8c8d20837dd34bfffb7e8] [ERROR] 	at java.lang.Thread.run(Thread.java:748)
[0691d6906f8e0af00de8c8d20837dd34bfffb7e8] [ERROR] 
[0691d6906f8e0af00de8c8d20837dd34bfffb7e8] [ERROR] -> [Help 1]
[0691d6906f8e0af00de8c8d20837dd34bfffb7e8] [ERROR] 
[0691d6906f8e0af00de8c8d20837dd34bfffb7e8] [ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[0691d6906f8e0af00de8c8d20837dd34bfffb7e8] [ERROR] Re-run Maven using the -X switch to enable full debug logging.
```

IMO we might consider allocating more memory to tests....

**Comment by strokyl** [comment]
I had a little commit if you don't agree just remove it

**Comment by chibenwa** [comment]
:+1:

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/mail/model/impl/MessageParser.java [hunk 8]
```diff
@@ -102,32 +105,29 @@ public MessageParser() {
...
-                }
+    private Stream<MessageAttachment> listAttachments(Entity entity, Context context) {
+        if (isMultipart(entity)) {
+            return listAttachments((Multipart) entity.getBody(), Context.fromEntity(entity));
+        }
+        if (isAttachment(entity, context)) {
+            try {
+                return Stream.of(retrieveAttachment(entity));
+            } catch (IllegalStateException e) {
+                LOGGER.error("The attachment is not well-formed", e);
```
Comment (aduprat) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/mail/model/impl/MessageParser.java]: Maybe just a warning, as the process is not failing.

File: mailbox/store/src/test/resources/eml/signedMessage.eml [hunk 9]
```diff
@@ -0,0 +1,84 @@
+Return-Path: <client@linagora.com>
+From: Important Client <client@linagora.com>
+To: David DOLCIMASCOLO <big.boss@linagora.com>, Michael BAILLY <big.big.boss@linagora.com>
```
Comment (aduprat) [inline mailbox/store/src/test/resources/eml/signedMessage.eml]: Please anonymize.

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/mail/model/impl/MessageParser.java [hunk 10]
```diff
@@ -220,7 +219,8 @@ private boolean isTextPart(Entity part) {
         return false;
     }
 
-    private byte[] getBytes(MessageWriter messageWriter, Body body) throws IOException {
+    private byte[] getBytes(Body body) throws IOException {
+        DefaultMessageWriter messageWriter = new DefaultMessageWriter();
```
[the remark this answers is not in the document]
Comment (mbaechler) [reply mailbox/store/src/main/java/org/apache/james/mailbox/store/mail/model/impl/MessageParser.java]: see https://github.com/linagora/james-project/pull/1155#discussion_r153427480

**Comment by chibenwa** [comment]
Removed the commit @mbaechler  was complaining about

**Pull request merged**